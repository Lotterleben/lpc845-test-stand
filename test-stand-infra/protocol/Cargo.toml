[package]
name    = "protocol"
version = "0.1.0"
authors = ["Hanno Braun <hanno@braun-embedded.com>"]
edition = "2018"

[dependencies.serde]
version          = "1.0.115"
default-features = false
features         = ["derive"]

[dependencies.lpc8xx-hal]
#git      = "https://github.com/lpc-rs/lpc8xx-hal.git"
path     = "../../../lpc8xx-hal"
features = ["845"]
