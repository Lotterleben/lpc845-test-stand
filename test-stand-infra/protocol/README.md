# protocol

Generic (i.e. independent of any specific test stand) building blocks that can be used to build a test-stand specific protocol. This crate is a dependency of the `messages` crate in this same repository.

See [top-level README](https://github.com/braun-embedded/lpc845-test-stand/blob/master/README.md) for more information.
