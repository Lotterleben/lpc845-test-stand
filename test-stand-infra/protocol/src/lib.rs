#![no_std]


pub mod pin;


use core::convert::TryFrom;

use serde::{
    Deserialize,
    Serialize,
};


/// A message from the test suite on the host to the test assistant
#[derive(Debug, Deserialize, Serialize)]
pub enum HostToAssistant<'r> {
    /// Instruct the assistant to send data to the target via USART
    SendUsart {
        mode: UsartMode,
        data: &'r [u8],
    },

    /// Instruct the assistant to change level of the target's input pin
    SetPin(pin::SetLevel<OutputPin>),

    /// Ask the assistant for the current level of a pin
    ReadPin(pin::ReadLevel<InputPin>),
}

impl From<pin::SetLevel<OutputPin>> for HostToAssistant<'_> {
    fn from(set_level: pin::SetLevel<OutputPin>) -> Self {
        Self::SetPin(set_level)
    }
}

impl From<pin::ReadLevel<InputPin>> for HostToAssistant<'_> {
    fn from(read_level: pin::ReadLevel<InputPin>) -> Self {
        Self::ReadPin(read_level)
    }
}


/// A message from the test assistant to the test suite on the host
#[derive(Debug, Deserialize, Serialize)]
pub enum AssistantToHost<'r> {
    /// Notify the host that data has been received from the target via USART
    UsartReceive {
        mode: UsartMode,
        data: &'r [u8],
    },

    /// Notify the host that the level of a pin has changed
    ReadPinResult(Option<pin::ReadLevelResult<InputPin>>),
}

impl<'r> TryFrom<AssistantToHost<'r>> for pin::ReadLevelResult<InputPin> {
    type Error = AssistantToHost<'r>;

    fn try_from(value: AssistantToHost<'r>) -> Result<Self, Self::Error> {
        match value {
            AssistantToHost::ReadPinResult(Some(result)) => {
                Ok(result)
            }
            _ => {
                Err(value)
            }
        }
    }
}


/// Specifies which mode a USART transmission uses
#[derive(Clone, Copy, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub enum UsartMode {
    Regular,
    Dma,
    FlowControl,
    Sync,
}


/// Represents one of the pins that the assistant is monitoring
#[derive(Clone, Copy, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub enum InputPin {
    Blue  = 0,
    Green = 1,
    Rts   = 2,
}

/// Represents one of the pins that the assistant can set
#[derive(Clone, Copy, Debug, Deserialize, Serialize, Eq, PartialEq)]
pub enum OutputPin {
    Pin5,
    Cts,
    Red,
}
