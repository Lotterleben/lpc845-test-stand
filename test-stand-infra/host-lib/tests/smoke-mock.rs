use host_lib::{
    mock::Mock,
    assistant::Assistant,
    conn::Conn,
};
use protocol::HostToAssistant;
use postcard;

use insta::assert_debug_snapshot;

#[test]
fn make_instance() {
    let mock = Mock::new();
    let test_hdl = mock.clone();

    let conn = Conn::from_serial_port(Box::new(mock)).unwrap();
    let _assistant = Assistant::new(conn);

    assert!(test_hdl.is_totally_empty());
}

#[test]
fn set_pin_5_high() {
    let mock = Mock::new();
    let test_hdl = mock.clone();

    let conn = Conn::from_serial_port(Box::new(mock)).unwrap();
    let mut assistant = Assistant::new(conn);
    assistant.set_pin_5_high().unwrap();

    let msg = test_hdl.pop_host_lib_data().unwrap();
    let mut msg_clone = msg.clone();
    let deser_msg: HostToAssistant = postcard::from_bytes_cobs(&mut msg_clone).unwrap();

    // TODO: Use the protocol to deserialize?
    // Add insta for snapshot tests
    assert_debug_snapshot!(
        "set_pin_5_high - bytes",
        &msg
    );

    assert_debug_snapshot!(
        "set_pin_5_high - parsed",
        &deser_msg
    );

    assert!(test_hdl.is_totally_empty());
}
