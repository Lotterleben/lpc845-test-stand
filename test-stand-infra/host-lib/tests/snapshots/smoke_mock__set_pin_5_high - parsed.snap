---
source: tests/smoke-mock.rs
expression: "&deser_msg"

---
SetPin(
    SetLevel {
        pin: Pin5,
        level: High,
    },
)
